### ISSUE TITLE
Allow MinGW windows compile of Tidy.

### ISSUE THREAD

**Issue by geoffmcl** [htacg/tidy-html5#349]
Started some of this conversation in #347, but moved it here under it own topic.

Since Nov 2015 I have been regularly compiling Tidy using [MinGW](http://www.mingw.org/) - "Minimalist GNU for Windows". Thanks to @d1vanov, and issue #206 for this... back in April 2015...

The aim would be to eventually also offer MinGW built binaries as an alternative to the current [MSVC10](http://microsoft-visual-cpp-express.soft32.com/). Actually I use the PRO version so I can build 32 and 64 bit versions, but maybe this can also be done with the `express` version.

It turns out remarkably easy after installing MinGW, and adding the path to its `bin` folder to the `PATH` env. -

```
$ cd build\cmake
$ cmake ..\.. -G "MinGW Makefiles" [options]
$ mingw32-make
```

I am still exploring MinGW, but the important thing to understand is that it uses a unix like setup, and a WIN32 port of gcc as the compiler, plus many other tools, and utilities, and many library ports, like even libws2_32.a, over 180 static libraries, and about 20 dlls... 

**All** standard windows includes are replaced with a MinGW set of includes - over 400 of them... These still define `WIN32` and `_WIN32`, but adds a `__MINGW32__` define for code macros. So if you wanted only say **native windows** code then you could use `#if defined(WIN32) && !defined(__MINGW32__)` macro...

Similarly cmake also still defines `WIN32`, but adds a `MINGW` define, so you can use combinations to get what you want. Like say `if (WIN32 AND NOT MINGW AND NOT MSYS)`, or `if (MSVC)`, etc, if you just wanted Windows without the unix/mingw/gcc mix in the build...

This MinGW build is broken in the current `localize_flat` branch!

```
libtidys.a(win_vsnprintf.c.obj):win_vsnprintf.c:(.text+0x1a): undefined reference to `_vscprintf_p'
libtidys.a(win_vsnprintf.c.obj):win_vsnprintf.c:(.text+0x5c): undefined reference to `_vsprintf_p'
```

Given that MinGW is a "Minimalist GNU for Windows" I thought this could be solved by adding the above **native windows** macro to these new sources, `win_vsnprintf.c|h`. That got rid of the above two errors, but introduced others.

Adding that macro to `tmbstr.c` got that to compile -

```
diff --git a/src/tmbstr.c b/src/tmbstr.c
index 5fcf515..2057b54 100644
--- a/src/tmbstr.c
+++ b/src/tmbstr.c
@@ -8,7 +8,7 @@
 #include "forward.h"
 #include "tmbstr.h"
 #include "lexer.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 #include "win_vsnprintf.h"
 #endif
 
```

And started to add it to `tidy.c` -

```
diff --git a/console/tidy.c b/console/tidy.c
index 49a77e2..ef00bcd 100644
--- a/console/tidy.c
+++ b/console/tidy.c
@@ -11,7 +11,7 @@
 #include "tidy.h"
 #include "language.h"
 #include "locale.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 /* Windows requires special help for printf positional format specifiers. */
 #include "win_vsnprintf.h"
 #define nest_(x) TY_(x)

```

And then language.c
```
diff --git a/src/language.c b/src/language.c
index 2635f2e..debf6d6 100644
--- a/src/language.c
+++ b/src/language.c
@@ -17,7 +17,7 @@
 #endif
 #include "tmbstr.h"
 #include "locale.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 #include "win_vsnprintf.h"
 #define nest_(x) TY_(x)
 #define printf nest_(win_printf)
```

But there are still errors compiling and linking tidy -

```
F:\Projects\tidy-local\src\language.c: In function 'tidyNormalizedLocaleName':
F:\Projects\tidy-local\src\language.c:376:4: warning: null argument where non-null required (argument 2) [-Wnonnull]
    strncpy( result + i, strlen( search ) >= 5 ? "_" : '\0', 1 );
    ^
Linking C static library libtidys.a
[ 49%] Built target tidy-static
Linking C executable tidy.exe
CMakeFiles\tidy.dir/objects.a(tidy.c.obj):tidy.c:(.text+0x43b): undefined reference to `prvTidywin_vsnprintf'
CMakeFiles\tidy.dir/objects.a(tidy.c.obj):tidy.c:(.text+0x483): undefined reference to `prvTidywin_vsnprintf'
libtidys.a(tmbstr.c.obj):tmbstr.c:(.text+0x512): undefined reference to `prvTidywin_vsnprintf'
collect2.exe: error: ld returned 1 exit status
```

So it seems I still have not got rid of the need for this special additional windows function in `tmbstr.c`, and `tidy.c`!

@d1vanov, maybe you could help with this? You just need to do a fresh `git pull`, and then `git checkout localize_flat` to get the source...

This might also apply to a `CYGWIN` build of tidy, which is another unix like setup in windows, but I do not have it installed anywhere... maybe by expanding a new WINDOWS_NATIVE macro with `defined(__CYGWIN__)`, something like -

```
#if (defined(WIN32) && !defined(__MINGW32__) && !defined(__CYGWIN__)
#define WINDOWS_NATIVE 1
#else
#define WINDOWS_NATIVE 0
#endif
```

Maybe something like that could be added to `tidyplatform.h`, with suitable adjustments of the `PLATFORM_NAME`?

While this is not a show-stopper for the merge #347, it would be great if this could be fixed before the merge... thanks...

**Comment by geoffmcl** [comment]
Below is a **full** patch to get it compiled using MinGW, but that turned up a **bug** in the string handling -

```
$ .\tidy -v
HTML Tidy for $s version $s
```

How come the platform and version are not shown? This works in native windows, and linux. Why not with MinGW???

Anyway, here is the full patch to get it compiled with WinGW

```
diff --git a/console/tidy.c b/console/tidy.c
index 49a77e2..ef00bcd 100644
--- a/console/tidy.c
+++ b/console/tidy.c
@@ -11,7 +11,7 @@
 #include "tidy.h"
 #include "language.h"
 #include "locale.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 /* Windows requires special help for printf positional format specifiers. */
 #include "win_vsnprintf.h"
 #define nest_(x) TY_(x)
diff --git a/include/tidyplatform.h b/include/tidyplatform.h
index 908692e..570a94e 100644
--- a/include/tidyplatform.h
+++ b/include/tidyplatform.h
@@ -450,7 +450,9 @@ extern "C" {
 #define utimbuf _utimbuf /* Windows seems to want utimbuf */
 #define stat _stat
 #define utime _utime
+#if !defined(__MINGW32__)
 #define vsnprintf prvTidywin_vsnprintf
+#endif
 #endif /* _WIN32 */
 
 #endif /* PRESERVE_FILE_TIMES */
diff --git a/src/language.c b/src/language.c
index 2635f2e..debf6d6 100644
--- a/src/language.c
+++ b/src/language.c
@@ -17,7 +17,7 @@
 #endif
 #include "tmbstr.h"
 #include "locale.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 #include "win_vsnprintf.h"
 #define nest_(x) TY_(x)
 #define printf nest_(win_printf)
diff --git a/src/tmbstr.c b/src/tmbstr.c
index 5fcf515..2057b54 100644
--- a/src/tmbstr.c
+++ b/src/tmbstr.c
@@ -8,7 +8,7 @@
 #include "forward.h"
 #include "tmbstr.h"
 #include "lexer.h"
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 #include "win_vsnprintf.h"
 #endif
 
diff --git a/src/win_vsnprintf.c b/src/win_vsnprintf.c
index 392e498..0ebb709 100644
--- a/src/win_vsnprintf.c
+++ b/src/win_vsnprintf.c
@@ -10,7 +10,7 @@
 #include "forward.h"
 #include "win_vsnprintf.h"
 
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 
 static char* TY_(new_string)( const ctmbstr fmt, va_list args );
 
diff --git a/src/win_vsnprintf.h b/src/win_vsnprintf.h
index 714cc5f..371494d 100644
--- a/src/win_vsnprintf.h
+++ b/src/win_vsnprintf.h
@@ -9,7 +9,7 @@
  * See tidy.h and access.h for the copyright notice.
  */
 
-#if defined(_WIN32)
+#if defined(_WIN32) && !defined(__MINGW32__)
 
 #include <stdio.h>
 #include "tidyplatform.h"

```

I could push these if you want? Advise...

**Comment by geoffmcl** [comment]
Ok, maybe the above patch is **all** wrong! Sort of the reverse of what is needed...

I can now see in Windows you have replaced the `printf` with `TY_(win_printf)` because the format string returned is `HTML Tidy for %1$s version %2$s`, which I assume can not be handled by standard windows `printf`.

So you divert it to use `out_string = TY_(new_string)( s, args );`, which uses `len = TY_(win_vsnprintf)( result, 0, fmt, args );` to first get the length, then allocate a buffer, and again use `TY_(win_vsnprintf) to format the string... using native `_vscprintf_p`...

That `malloc`, which can fail, and thus you would do a hard exit(2), can be replaced with libtidy allocator, which in turn can be replaced with a user supplied allocator, allowing a library tidy users to avoid such a hard exit, but that is later... libtidy can only have one hard exit, and that is in its default allocator, but this can be overcome by using your own allocator and dealing with memory failures as the lib user wants... see #296

And I further assume that in linux `printf` can handle such position parameters like `%1$s`?

But this MinGW is in windows, and not linux, so it too would need to uses some substitute, since obviously it only prints out `HTML Tidy for $s version $s`!

But what substitute can MinGW offer in windows? As mentioned I do not think it has the native vsnprintf_p emulated in its libraries... could not find it...

So the above patch goes the wrong way... we are still in Windows... maybe missing some native functions in the MinGW emulation libraries... but we still need to swap out `printf` for something else... but what?

It seems a serious impediment that our current localization presently does not support a MinGW build! But as indicated, maybe there are work arounds that can be done...

But forget the above patch...

**Comment by balthisar** [comment]
@geoffmcl, thanks for looking into this. You can see I haven't been too available over the weekend, so I've not install MinGW in my Windows environment yet.

As I browse around I see some indication that MinGW's implementation is supposed to be POSIX compliant (i.e., work with the locational parameters). I found this bit:

>You can still use #define __USE_MINGW_ANSI_STDIO 1 to get the posix variant instead of the MS one.

But obviously until I install MinGW I can't test this, but I do see that macro used in several GitHub results that turn up on Google.

**Comment by geoffmcl** [comment]
@balthisar Keeping my patch, and adding the `#define __USE_MINGW_ANSI_STDIO 1`, which I added to `CMakleLists.txt` as -

```
if (MINGW)
    add_definitions ( -D__USE_MINGW_ANSI_STDIO=1 )
endif ()
```
certainly got a different result. Now the output is `HTML Tidy for %1$s version %2$s`! I am not sure that is an improvement or regression ;=)) And of course this is seen in all messages that have positional format params...

I can see that macro **is** used several times on the MinGW `<stdio.h>` header... so makes a difference...

Now it seems the positional parameters, which I have never tried before, and still understanding, are completely ignored in the code `__retval = __mingw_vprintf( __format, __local_argv );` seen in the gdb debugger? 

So it seems from other tests `printf` supports them in native linux, but neither it, which would be native WIN32 `printf`, nor this substitute `_mingw_vprintf` support them in MinGW! The joys of cross compiling...

But I am sure there is a solution, and will keep searching...

**Comment by balthisar** [comment]
Thanks, @geoffmcl. I'll try to look into it when I get near a Windows machine or VM again. For the time being (probably until the weekend) my only dev system is my Mac.

**Comment by balthisar** [comment]
Oops, I was in the wrong thread. I see you've already stumbled upon `_mingw_printf` family, and it's not working. Will continue to look for a solution.

**Comment by geoffmcl** [comment]
@balthisar addressing also your comments in #347...

Re: Which MinGW?

Well my first install was from what seems like it main **canonical** home - http://www.mingw.org/ - and its SF downloads - https://sourceforge.net/projects/mingw/files/ - downloaded mingw-get-setup.exe, and installed in in D:\MinGW. And use a set-mingw.bat to add MinGW to the environment of that specific console/terminal before the build... in part -

```
@set MINGW_ROOT=D:\MinGW
@if NOT EXIST %MINGW_ROOT%\nul goto NOROOT
@set PATH=%MINGW_ROOT%\bin;%PATH%
@set INCLUDE=%MINGW_ROOT%\include
@set LIB=%MINGW_ROOT%\lib
```

As indicated, this seems a very full 32-bit port, with over 400 headers, 200 libraries, lots of tools, utilities... so seemed to offer the best chance, especially when you pointed out the define __USE_MINGW_ANSI_STDIO which, through its own `<stdio.h>`, converts `printf` to `_mingw_printf` so looked good, but unfortunately does **not** appear to support positional parameters!!!

I have now installed MinGW-w64, but this has just a measly 22 headers, and `<stdio.h>` is **not** one of them, and only has about 30 libraries. So this package, from https://sourceforge.net/projects/mingw-w64/ seems to try to mostly use the native windows headers, and libraries... so that means `printf` is not converted, and thus the tidy.exe built is back to the first output `HTML Tidy for $s version $s`...

So maybe for the MinGW-w64 build maybe I need to **reverse** my patch, and hopefully be able to use the native windows `_vscprintf_p`??? Will try to experiment with that... 

But for now would like to at least get the http://www.mingw.org/ 32-bit working...

And concerning the `__MINGW32__` macro, I guess this could be like Windows in that respect. `WIN32` and `_WIN32` are defined in both the 32 and 64 bit builds. For the 64-bit build, there is an additional `WIN64` that can be defined to move more things into 64-bits. And this post - http://mingw.5.n7.nabble.com/Macros-MINGW32-AND-MINGW64-td26319.html - suggests **both** `__MINGW32__` **and** `__MINGW64__` are defined in the MinGW 64-bit build. Will try to test that...

Re: POSIX

To quote directly from the - http://www.mingw.org/ - site **"MinGW, being Minimalist, does not, and never will, attempt to provide a POSIX runtime environment for POSIX application deployment on MS-Windows. If you want POSIX application deployment on this platform, please consider Cygwin instead."** 

I suppose I also need to install CYGWIN somewhere, and see what it does... I did have it installed a long time back, but decided I did not like the special proprietary DLL it embeds in every application! But just to see what it does, uses...

Re: Positional Format Parameters

All this is about supported such `positional parameters`, which I must admit I had never heard of before a few days ago. I have **never** seen them in any of the **many** cross-ported projects I have worked on. Even the one project I remember with PO languages files, FGRUN, does not use them!

Are these really needed in tidy's case. The bulk of the warning and error messages are deliberately short. Would there ever be a case for putting the parameters in a different order in the final string? It seems not. I note they have been used in tidy.pot even when there is only one parameter, like `"Can't open \"%1$s\"\n"`!!!

And the largest is `%4$`, in `"%1$u %2$s, %3$u %4$s were found!"`, but you can not mess with that too much because 2 are ints, and 2 are strings, by why would you need to **`mess`** with it?

Yes languages like french put say some adjective in reverse, like a `the blue bag` becomes a `le sac bleu`, but it seems this type of difference can be handled in the translation strings, and I can not see how this would be related to using such positional parameters...

If we could get rid of those, then this problem goes away ;=)) Are they really worth the pain and effort???

**Comment by balthisar** [comment]
> If we could get rid of those, then this problem goes away ;=)) Are they really worth the pain and effort???

A damned good question!

I've been looking at the minGW standard branch. I've simplified the current branch for testing:

~~~
#if !defined(__MINGW32__)
		return _vscprintf_p(fmt, ap);
#else
		return __mingw_vsnprintf(s, 0, fmt, ap);
#endif
~~~

...just to deal with a single file. Obviously looking for a substitution to the `__mingw_vsnprintf` bit. Trying custom spec files and using newer Windows C runtime DLLs and so on an so forth, just to include what's already native in Windows! I don't know if this is a MinGW flaw or a feature.

I've also look at implementations of C99 printf family, but they don't seem to implement POSIX strings either.

I'm going to give it one last go, and then just eliminate the positional parameters if I don't find an expeditious solution. At least we can push localize, and make the positional parameters a future enhancement if someone with some strange language requests them. Can you tell I'm tired? ;-)

**Comment by gagern** [comment]
See http://sourceforge.net/p/mingw/mingw-org-wsl/ci/master/tree/src/libcrt/stdio/ and http://sourceforge.net/p/mingw/mingw-org-wsl/ci/master/tree/src/libcrt/include/pformat.h for the MinGW runtime implementation of these functions. These files seem to be in place since 2012.

**Comment by balthisar** [comment]
@gagern, I looked at these yesterday, specifically [pformat.c](http://sourceforge.net/p/mingw/mingw-org-wsl/ci/21762bb4a1bd0c88c38eead03f59e8d994349e83/tree/misc/src/libcrt/stdio/pformat.c#l153) and I'm not seeing that there's support for positional parameters.

**Comment by balthisar** [comment]
I'm too inconsistent with where I keep putting the discussion. Head to #347: We can build with WinGW now, including positional parameter support. I think it still needs some help for a "friendly" build solution, but it's working now both with MSVC and MingW.

**Comment by geoffmcl** [comment]
@balthisar absolutely **bloody** amazing - please excuse my french ;=))

Yes, after following your README.txt, and modifying my D:\MinGW install with the 2 new files, my MinGW generated tidy.exe now correctly emits `HTML Tidy for Windows version 5.1.33localizing` - PHEW! That is **GREAT**.

I immediately wanted to help you file a PR to MinGW. I did the first step - forked their SF repo - https://sourceforge.net/u/geoffmc/mingw-fork/ci/master/tree/ - and will explore this further... I agree adding these files to tidy source, and presenting quite complex build instructions is not the best way to go... **but it works**!

And that is asside from the fact that I prefer to **not** add a `mingw` folder. I wanted to remove `cmake` and `win64` folders... I think `build` should be for what ever available tools you use to build, but never got around to that...

What an amazing tool they offer, the `dlltool`. To be able to extract modules from native windows DLLS and create a linkable static library is something I have searched for in the past. You are a **super sleuth** in finding this!!! And using the MSVC `dumpbin` to explore... I am sure I would never have got there...

BTW, `dlltool` offers an option I have always wanted for tidy - `@<file> - Read options from <file>`... but that is another story...

And to then know to add a `specs` file to automate that linkage... If we can find the `msvcrt.def` they use to create the existing distributed `libmsvcrt.a`, maybe we would not need to add a `libmsvcrt_tidy.a`... That old, circa 2011 msvcrt.dll exports what?, over a thousand functions... what else did they not add to theirs?

So, whether we really need positional format parameters or **not**, and thus have that additional WIN32 only source module, now they are available... and that is another small cmake fix I would like - only add that source to the link if in WIN32... no need to add a blank object to the linux link...

Anyway, out of time tonight, but lots to explore ;=)) thanks...

**Comment by balthisar** [comment]
@geoff, I know we didn't want to add things to `/build` but it seemed like the most expedient thing to do for now. I'm hopeful we can get the MinGW people to update the repo so this can be a completely native build without the kludges. It looks like the last commits were in 2013, though, and so I wonder if the project is withering from lack of interest, too.

A lot of my research tried to give me answers from MinGW-64, which _does_ appear to still be active and fully supported. Apparently it was forked from MinGW some years ago, and there seems to be some hard feelings between the groups that don't bespeak a unification, but maybe I'll give it a tryout, too, anyway.

As for CMake, yeah, I agree, no need to add blank objects. I'm not the CMake guru you are. Assuming we don't get the MinGW source fixed, I imagine that CMake would require special detection of MinGW, such as when the -G option is used.

I don't know if these are possible:
- Add a processing step to build the `libmsvcrt_tidy.dll` from the `def` automatically.
- Provide the additional gcc linking parameter to `libmsvcrt_tidy.dll`.

Accomplishing the first one would avoid distributing a binary as part of the source, and accomplishing the second one would avoid having to install a custom gcc specs file just to build tidy.

Obviously would be best if we can have MinGW update their import libraries instead!

**Comment by balthisar** [comment]
> And to then know to add a specs file to automate that linkage... If we can find the msvcrt.def they use to create the existing distributed libmsvcrt.a, maybe we would not need to add a libmsvcrt_tidy.a... That old, circa 2011 msvcrt.dll exports what?, over a thousand functions... what else did they not add to theirs?

The patch looks relatively simple. I've not experimented, but it seems that adding the required functions to `mingw/lib/lib32/msvcrt.def.in` and adding them to `stdio.h` is all it would take to make this happen. The Makefile looks like it builds the actual import library DLLs.

I posted a [ticket](https://sourceforge.net/p/mingw/bugs/2280/) on their SF bug tracker asking for guidance on submitting the patch. I know they probably tend to get a lot of dumb questions giving them the appearance of being temperamental, so I took a cautious approach with their possible acceptance of a patch. Looking at the DEF file there are some conditionals that are handled in the Makefile, and since I don't know exactly at what point these required functions were added to `msvcrt.dll` we may have to curtain them off -- **if** the MinGW folks are okay with this.
